onboarding/mlsys/1-gradients.md

```diff
@@ -0,0 +1,493 @@
+if the slope is really small we think we're getting close to the slope and we
+only take a small step.
+
+So let's define the step size `mstep` as:
+
+    mstep = - gradient * lr
+
+But where does that `gradient` term come from?
+
+## The Gradient
+
+In the case of our simple model, we can use calculus to calculate the slope of
+our loss when plotted against `m` (technically, the "derivative of our loss
+with respect to `m`").  In the case of a one-variable model, this is identical
+to the gradient of the loss.
+
+We'll start by recalling the definition of `loss_point` in terms of `m`:
+
+    loss_point = (ytrue - (m*x))**2 / N
+
+Then apply the product rule and the chain rule from to get the derivative with
+respect to the weight `m`:
+
+    dloss_point/dm = - 2*x*(ytrue - m*x) / N
+
+And since our total `loss_mse` is just the sum of several `loss_point` values,
+we know from calculus that the derivative of the total `loss_mse` is the sum
+of the per-point derivatives:
+
+    dloss_mse/dm = - 2 * ( x1*(ytrue1 - m*x1) + ... + xN*(ytrueN - m*xN) ) / N
+
+### Q5:
+
+Write a function to calculate the gradient for a single data point.  It should
+take `m`, a single `(x, ytrue)` pair, and `N` as inputs and return
+`dloss_point/dm`.
+
+You can test with:
+
+```python
+assert grad_point(m=2, data=(1, 1), N=4) == 0.5
+assert grad_point(m=2, data=(2, 2), N=4) == 2.0
+```
+
+Then write a function to calculate the total gradient for a list of `(x,
+ytrue)` pairs.  It should take `m` and a list of `(x, ytrue)` pairs as inputs.
+
+You can test with:
+
+```python
+assert grad_multi(m=2, data=[(1, 1), (2, 2), (3, 3), (4, 4)]) == 15.0
+```
+
+Is it mathematically valid to call the first function from the second function
+to simplify writing the second?  Why or why not?
+
+## Training by Gradient Descent
+
+The basic gradient descent loop algorithm is:
+
+1. Pick an initial model weight.
+2. Select one or more `(x, ytrue)` pairs from the dataset.
+3. Calculate a prediction for the selected inputs with a label (Q2).
+4. Calculate the gradient using the prediction vs label (Q3).
+5. Subtract `gradient * lr` from our model weight ("Minimizing the loss").
+6. Repeat steps 2—5 until you decide to stop.
+
+### Q6
+
+Write a training loop to do gradient descent on our `ypred = m*x` model.  Your
+function should take the following inputs:
+
+- initial model weight `m`
+- a learning rate `lr` for calculating the step
+- a dataset, which is a list of `(x, ytrue)` data pairs
+- a number of iterations to do (one "iteration" is steps 2-5).
+
+Your function should return the final trained weight `m`.
+
+Additionally, in the "select one or more `(x, ytrue)` input pairs" step, for
+now you should just select the next `(x, ytrue)` pair in the dataset, and it
+should wrap back to the beginning of the dataset after reaching the end.  So if
+your dataset is of length 10, then your eleventh iteration should reuse the
+first `(x, ytrue)` pair of your dataset.
```

Comment:
Could we explain a quick rationale why we would make it work in this way? We get to it later because it leads into the concept of batching, but I think it would make sense to explain that we do this because in a real training there could be too many data points for us to be able to calculate every one of them on every step.

Reply:
sure